File: internal/agentrun/intelligence.go
package agentrun

import (
	"sort"
	"time"
)

// IntelligenceState is a conservative interpretation of observable evidence.
// It is descriptive only and never changes a p-track task or run lifecycle.
type IntelligenceState string

const (
	IntelligenceUnknown             IntelligenceState = "unknown"
	IntelligenceWorking             IntelligenceState = "working"
	IntelligenceWaiting             IntelligenceState = "waiting"
	IntelligenceBlocked             IntelligenceState = "blocked"
	IntelligenceCompleted           IntelligenceState = "completed"
	IntelligenceFailed              IntelligenceState = "failed"
	IntelligencePotentiallyDrifting IntelligenceState = "potentiallyDrifting"
)

type IntelligenceConfidence string

const (
	ConfidenceLow    IntelligenceConfidence = "low"
	ConfidenceMedium IntelligenceConfidence = "medium"
	ConfidenceHigh   IntelligenceConfidence = "high"
)

// IntelligenceEvidence exposes only stable evidence metadata, never event
// summaries, paths, subjects, provider payloads, terminal content, or errors.
type IntelligenceEvidence struct {
	EventID    string     `json:"eventId,omitempty"`
	Kind       EventKind  `json:"kind,omitempty"`
	Phase      EventPhase `json:"phase,omitempty"`
	ObservedAt time.Time  `json:"observedAt,omitempty"`
	Reason     string     `json:"reason"`
}

type RunIntelligence struct {
	RunID       string                 `json:"runId"`
	State       IntelligenceState      `json:"state"`
	Confidence  IntelligenceConfidence `json:"confidence"`
	Evidence    []IntelligenceEvidence `json:"evidence"`
	EventCount  int                    `json:"eventCount"`
	LastEventAt time.Time              `json:"lastEventAt,omitempty"`
}

// DeriveRunIntelligence reduces ordered structured evidence without treating
// silence as waiting/completion, an operation failure as a run failure, or an
// unrelated file/commit as drift. A successful process exit alone likewise
// does not claim that the agent completed its objective.
func DeriveRunIntelligence(run Run, events []Event) RunIntelligence {
	ordered := currentRunEvents(run, events)
	result := RunIntelligence{
		RunID:      run.ID,
		State:      IntelligenceUnknown,
		Evidence:   []IntelligenceEvidence{},
		EventCount: len(ordered),
	}
	if len(ordered) > 0 {
		result.LastEventAt = ordered[len(ordered)-1].ObservedAt
	}
	if run.Exit != nil && run.Exit.Code != 0 {
		return withRunEvidence(result, IntelligenceFailed, ConfidenceHigh,
			IntelligenceEvidence{Reason: "nonzero_process_exit"})
	}

	live := runIsActive(run)
	for index := len(ordered) - 1; index >= 0; index-- {
		event := ordered[index]
		evidence := eventIntelligenceEvidence(event, "explicit_event")
		switch {
		case event.Kind == EventLifecycle && event.Phase == EventFailed:
			evidence.Reason = "explicit_lifecycle_failure"
			return withRunEvidence(result, IntelligenceFailed, ConfidenceHigh, evidence)
		case event.Kind == EventLifecycle && event.Phase == EventCompleted:
			evidence.Reason = "explicit_lifecycle_completion"
			return withRunEvidence(result, IntelligenceCompleted, ConfidenceHigh, evidence)
		case event.Kind == EventError && fatalEventClass(event.ErrorClass):
			evidence.Reason = "explicit_fatal_error"
			return withRunEvidence(result, IntelligenceFailed, ConfidenceHigh, evidence)
		case event.Phase == EventBlocked && live:
			evidence.Reason = "explicit_blocked_event"
			return withRunEvidence(result, IntelligenceBlocked, ConfidenceMedium, evidence)
		case event.Phase == EventWaiting && live:
			evidence.Reason = "explicit_waiting_event"
			return withRunEvidence(result, IntelligenceWaiting, ConfidenceMedium, evidence)
		case live && driftEventClass(event.ErrorClass) && eventCorrelationIsCurrent(run, event):
			evidence.Reason = "explicit_scope_mismatch"
			return withRunEvidence(result, IntelligencePotentiallyDrifting, ConfidenceMedium, evidence)
		case live && (event.Phase == EventStarted || event.Phase == EventProgress ||
			event.Phase == EventCompleted || event.Phase == EventFailed):
			if event.Phase == EventFailed {
				evidence.Reason = "operation_failure_while_run_live"
			} else {
				evidence.Reason = "recent_observable_activity"
			}
			return withRunEvidence(result, IntelligenceWorking, ConfidenceMedium, evidence)
		}
	}
	if live {
		return withRunEvidence(result, IntelligenceWorking, ConfidenceLow,
			IntelligenceEvidence{Reason: "live_run_without_structured_progress"})
	}
	return result
}

func currentRunEvents(run Run, events []Event) []Event {
	ordered := make([]Event, 0, len(events))
	for _, event := range events {
		if event.ModelVersion != EventModelVersion || event.RunID != run.ID ||
			event.Provider != run.Provider || event.HostSequence == 0 ||
			event.LifecycleRevision != run.LifecycleRevision {
			continue
		}
		ordered = append(ordered, cloneEvent(event))
	}
	sort.SliceStable(ordered, func(i, j int) bool {
		if ordered[i].HostSequence != ordered[j].HostSequence {
			return ordered[i].HostSequence < ordered[j].HostSequence
		}
		return ordered[i].ObservedAt.Before(ordered[j].ObservedAt)
	})
	return ordered
}

func eventIntelligenceEvidence(event Event, reason string) IntelligenceEvidence {
	return IntelligenceEvidence{
		EventID:    event.ID,
		Kind:       event.Kind,
		Phase:      event.Phase,
		ObservedAt: event.ObservedAt,
		Reason:     reason,
	}
}

func withRunEvidence(
	result RunIntelligence,
	state IntelligenceState,
	confidence IntelligenceConfidence,
	evidence IntelligenceEvidence,
) RunIntelligence {
	result.State = state
	result.Confidence = confidence
	result.Evidence = []IntelligenceEvidence{evidence}
	return result
}

func fatalEventClass(class string) bool {
	switch class {
	case "fatal", "fatal_error", "session_failure", "process_failure":
		return true
	default:
		return false
	}
}

func driftEventClass(class string) bool {
	switch class {
	case "scope_mismatch", "task_mismatch", "repository_mismatch":
		return true
	default:
		return false
	}
}

func eventCorrelationIsCurrent(run Run, event Event) bool {
	current := run.Association
	correlation := event.Correlation
	return current != nil && correlation.TaskID != 0 &&
		current.ProjectRoot == run.ProjectRoot && current.LiveID == run.ID &&
		correlation.ProjectRoot == run.ProjectRoot &&
		correlation.TerminalID == run.TerminalID &&
		correlation.PlanID == current.Target.PlanID &&
		correlation.TaskID == current.Target.TaskID &&
		correlation.Generation == current.Generation &&
		correlation.AssociationRevision == current.Revision
}

File: crates/ptrack-agent/src/intelligence.rs
use serde::{Deserialize, Serialize};

use crate::run::run_is_active;
use crate::{EVENT_MODEL_VERSION, Event, EventKind, EventPhase, Run, Timestamp};

#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub enum IntelligenceState {
    #[default]
    #[serde(rename = "unknown")]
    Unknown,
    #[serde(rename = "working")]
    Working,
    #[serde(rename = "waiting")]
    Waiting,
    #[serde(rename = "blocked")]
    Blocked,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "failed")]
    Failed,
    #[serde(rename = "potentiallyDrifting")]
    PotentiallyDrifting,
}

impl IntelligenceState {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Unknown => "unknown",
            Self::Working => "working",
            Self::Waiting => "waiting",
            Self::Blocked => "blocked",
            Self::Completed => "completed",
            Self::Failed => "failed",
            Self::PotentiallyDrifting => "potentiallyDrifting",
        }
    }
}

#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub enum IntelligenceConfidence {
    #[default]
    #[serde(rename = "")]
    Unset,
    #[serde(rename = "low")]
    Low,
    #[serde(rename = "medium")]
    Medium,
    #[serde(rename = "high")]
    High,
}

impl IntelligenceConfidence {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Unset => "",
            Self::Low => "low",
            Self::Medium => "medium",
            Self::High => "high",
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IntelligenceEvidence {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub event_id: String,
    #[serde(default, skip_serializing_if = "is_unset_kind")]
    pub kind: EventKind,
    #[serde(default, skip_serializing_if = "is_unset_phase")]
    pub phase: EventPhase,
    #[serde(default, skip_serializing_if = "is_zero_timestamp")]
    pub observed_at: Timestamp,
    pub reason: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RunIntelligence {
    pub run_id: String,
    pub state: IntelligenceState,
    pub confidence: IntelligenceConfidence,
    pub evidence: Vec<IntelligenceEvidence>,
    pub event_count: usize,
    #[serde(default, skip_serializing_if = "is_zero_timestamp")]
    pub last_event_at: Timestamp,
}

#[must_use]
pub fn derive_run_intelligence(run: &Run, events: &[Event]) -> RunIntelligence {
    let ordered = current_run_events(run, events);
    let mut result = RunIntelligence {
        run_id: run.id.clone(),
        state: IntelligenceState::Unknown,
        confidence: IntelligenceConfidence::Unset,
        evidence: Vec::new(),
        event_count: ordered.len(),
        last_event_at: ordered
            .last()
            .map_or(Timestamp::ZERO, |event| event.observed_at),
    };
    if run.exit.as_ref().is_some_and(|exit| exit.code != 0) {
        return with_evidence(
            result,
            IntelligenceState::Failed,
            IntelligenceConfidence::High,
            IntelligenceEvidence::reason("nonzero_process_exit"),
        );
    }
    let live = run_is_active(run);
    for event in ordered.iter().rev() {
        let mut evidence = IntelligenceEvidence::event(event, "explicit_event");
        let decision = if event.kind == EventKind::Lifecycle && event.phase == EventPhase::Failed {
            "explicit_lifecycle_failure".clone_into(&mut evidence.reason);
            Some((IntelligenceState::Failed, IntelligenceConfidence::High))
        } else if event.kind == EventKind::Lifecycle && event.phase == EventPhase::Completed {
            "explicit_lifecycle_completion".clone_into(&mut evidence.reason);
            Some((IntelligenceState::Completed, IntelligenceConfidence::High))
        } else if event.kind == EventKind::Error && fatal_event_class(&event.error_class) {
            "explicit_fatal_error".clone_into(&mut evidence.reason);
            Some((IntelligenceState::Failed, IntelligenceConfidence::High))
        } else if event.phase == EventPhase::Blocked && live {
            "explicit_blocked_event".clone_into(&mut evidence.reason);
            Some((IntelligenceState::Blocked, IntelligenceConfidence::Medium))
        } else if event.phase == EventPhase::Waiting && live {
            "explicit_waiting_event".clone_into(&mut evidence.reason);
            Some((IntelligenceState::Waiting, IntelligenceConfidence::Medium))
        } else if live
            && drift_event_class(&event.error_class)
            && event_correlation_is_current(run, event)
        {
            "explicit_scope_mismatch".clone_into(&mut evidence.reason);
            Some((
                IntelligenceState::PotentiallyDrifting,
                IntelligenceConfidence::Medium,
            ))
        } else if live
            && matches!(
                event.phase,
                EventPhase::Started
                    | EventPhase::Progress
                    | EventPhase::Completed
                    | EventPhase::Failed
            )
        {
            if event.phase == EventPhase::Failed {
                "operation_failure_while_run_live"
            } else {
                "recent_observable_activity"
            }
            .clone_into(&mut evidence.reason);
            Some((IntelligenceState::Working, IntelligenceConfidence::Medium))
        } else {
            None
        };
        if let Some((state, confidence)) = decision {
            return with_evidence(result, state, confidence, evidence);
        }
    }
    if live {
        result = with_evidence(
            result,
            IntelligenceState::Working,
            IntelligenceConfidence::Low,
            IntelligenceEvidence::reason("live_run_without_structured_progress"),
        );
    }
    result
}

pub(crate) fn current_run_events(run: &Run, events: &[Event]) -> Vec<Event> {
    let mut ordered: Vec<Event> = events
        .iter()
        .filter(|event| {
            event.model_version == EVENT_MODEL_VERSION
                && event.run_id == run.id
                && event.provider == run.provider
                && event.host_sequence != 0
                && event.lifecycle_revision == run.lifecycle_revision
        })
        .cloned()
        .collect();
    ordered.sort_by(|left, right| {
        left.host_sequence
            .cmp(&right.host_sequence)
            .then(left.observed_at.cmp(&right.observed_at))
    });
    ordered
}

impl IntelligenceEvidence {
    fn reason(reason: &str) -> Self {
        Self {
            event_id: String::new(),
            kind: EventKind::Unset,
            phase: EventPhase::Unset,
            observed_at: Timestamp::ZERO,
            reason: reason.to_owned(),
        }
    }

    fn event(event: &Event, reason: &str) -> Self {
        Self {
            event_id: event.id.clone(),
            kind: event.kind,
            phase: event.phase,
            observed_at: event.observed_at,
            reason: reason.to_owned(),
        }
    }
}

fn with_evidence(
    mut result: RunIntelligence,
    state: IntelligenceState,
    confidence: IntelligenceConfidence,
    evidence: IntelligenceEvidence,
) -> RunIntelligence {
    result.state = state;
    result.confidence = confidence;
    result.evidence = vec![evidence];
    result
}

fn fatal_event_class(value: &str) -> bool {
    matches!(
        value,
        "fatal" | "fatal_error" | "session_failure" | "process_failure"
    )
}

fn drift_event_class(value: &str) -> bool {
    matches!(
        value,
        "scope_mismatch" | "task_mismatch" | "repository_mismatch"
    )
}

fn event_correlation_is_current(run: &Run, event: &Event) -> bool {
    let Some(current) = run.association.as_ref() else {
        return false;
    };
    let correlation = &event.correlation;
    correlation.task_id != 0
        && current.project_root == run.project_root
        && current.live_id == run.id
        && correlation.project_root == run.project_root
        && correlation.terminal_id == run.terminal_id
        && correlation.plan_id == current.target.plan_id
        && correlation.task_id == current.target.task_id
        && correlation.generation == current.generation
        && correlation.association_revision == current.revision
}

const fn is_unset_kind(value: &EventKind) -> bool {
    matches!(value, EventKind::Unset)
}
const fn is_unset_phase(value: &EventPhase) -> bool {
    matches!(value, EventPhase::Unset)
}
const fn is_zero_timestamp(value: &Timestamp) -> bool {
    value.is_zero()
}
